creator-node/src/services/stateMachineManager/makeOnCompleteCallback.js

```diff
@@ -0,0 +1,186 @@
+ * a job that successfully completed and read its output for new jobs that will be enqueued in bulk as well
+ * as metrics that will be recorded.
+ * The expected syntax that a job result must output in order to for this function to enqueue more
+ * jobs upon completion is:
+ * {
+ *   jobsToEnqueue: {
+ *     [QUEUE_NAMES.STATE_MONITORING]: [
+ *       {
+ *         jobName: <any job name from JOB_NAMES constant>,
+ *         jobData: <object containing data that this job type expects>
+ *       },
+ *       ...
+ *     ],
+ *     [QUEUE_NAMES.STATE_RECONCILIATION]: [
+ *       {
+ *         jobName: <any job name from JOB_NAMES constant>,
+ *         jobData: <object containing data that this job type expects>
+ *       },
+ *       ...
+ *     ]
+ *   }
+ * }
+ * @dev MUST be bound to a class containing an `enabledReconfigModes` property.
+ *      See usage in index.js (in same directory) for example of how it's bound to StateMachineManager.
+ *
+ * @param {BullQueue} monitoringQueue the queue that handles state monitoring jobs
+ * @param {BullQueue} reconciliationQueue the queue that handles state reconciliation jobs
+ * @param {Object} prometheusRegistry the registry of prometheus metrics
+ * @returns a function that:
+ * - takes a jobId (string) and result (string) of a job that successfully completed
+ * - parses the result (string) into JSON
+ * - bulk-enqueues all jobs under result[QUEUE_NAMES.STATE_MONITORING] into the state monitoring queue
+ * - bulk-enqueues all jobs under result[QUEUE_NAMES.STATE_RECONCILIATION] into the state reconciliation queue
+ */
+module.exports = function (
+  monitoringQueue,
+  reconciliationQueue,
+  prometheusRegistry
+) {
+  return async function (jobId, resultString) {
+    // Create a logger so that we can filter logs by the tag `jobId` = <id of the job that successfully completed>
+    const logger = createChildLogger(baseLogger, { jobId })
+
+    // update-replica-set jobs need enabledReconfigModes as an array.
+    // `this` comes from the function being bound via .bind() to ./index.js
+    if (!this?.hasOwnProperty('enabledReconfigModesSet')) {
+      logger.error(
+        'Function was supposed to be bound to StateMachineManager to access enabledReconfigModesSet! Update replica set jobs will not be able to process!'
+      )
+      return
+    }
+    const enabledReconfigModes = Array.from(this.enabledReconfigModesSet)
+
+    // Bull serializes the job result into redis, so we have to deserialize it into JSON
+    let jobResult = {}
+    try {
+      jobResult = JSON.parse(resultString) || {}
+    } catch (e) {
+      logger.warn(`Failed to parse job result string: ${resultString}`)
+      return
+    }
+
+    const { jobsToEnqueue, metricsToRecord } = jobResult
+
+    if (jobsToEnqueue) {
+      // Enqueue monitoring jobs
+      const monitoringJobs = jobsToEnqueue[QUEUE_NAMES.STATE_MONITORING] || []
+      await enqueueMonitoringJobs(monitoringJobs, monitoringQueue, logger)
```

Comment:
I prefer them to be called with an empty array because the extra `if` code doesn't help anything, and this way we get an extra log saying "attempted to enqueue 0 monitoring jobs", which lets us know that it read the output correctly

Reply:
fair